Lib/concurrent/futures/_base.py

```diff
@@ -170,6 +170,17 @@ def _create_and_install_waiters(fs, return_when):
 
     return waiter
 
+
+def _yield_future(fs, waiter, ref_collect=()):
+    while fs:
+        with fs[0]._condition:
```

Comment:
I we do not remove waiter, second result set on future will trigger waiter and cause KeyError, future is already returned and reference cleared.

For example:
```
>>> from concurrent.futures import Future, as_completed
>>> fs_finished = Future()
>>> fs = Future()
>>> fs_finished.set_result("")
>>> for x in as_completed([fs_finished, fs, Future()]):
...     fs.set_result(None)
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/grzegorz/cpython/Lib/concurrent/futures/_base.py", line 235, in as_completed
    pending.remove(future)
KeyError: <Future at 0x7f566ff6fc20 state=finished returned NoneType>
```
This error occurs in both version.

However docstring for `Future.set_result` and `Future.set_exception` says:
> Should only be used by Executor implementations and unit tests.

So maybe we should ignore this case?